src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Provider that caches semantic models for requested trees, with a strong reference to the model.
+    /// Clients using this provider are responsible for maintaining the lifetime of the entries in this cache,
+    /// and should invoke <see cref="ClearCache(SyntaxTree, Compilation)"/> and <see cref="ClearCache(Compilation)"/> to clear entries when appropriate.
+    /// For example, <see cref="CompilationWithAnalyzers"/> uses this provider to ensure that semantic model instances
+    /// are shared between the compiler and analyzers for improved analyzer execution performance. The underlying
+    /// <see cref="AnalyzerDriver"/> executing analyzers clears per-tree entries in the cache whenever a <see cref="CompilationUnitCompletedEvent"/>
+    /// has been processed, indicating all relevant analyzers have executed on the corresponding syntax tree for the event.
+    /// Similarly, it clears the entire compilation wide cache whenever a <see cref="CompilationCompletedEvent"/> has been processed,
+    /// indicating all relevant analyzers have executed on the entire compilation.
+    /// </summary>
+    internal sealed class CachingSemanticModelProvider : SemanticModelProvider
+    {
+        private readonly ConditionalWeakTable<Compilation, PerCompilationProvider> _providerCache;
+        private readonly ConditionalWeakTable<Compilation, PerCompilationProvider>.CreateValueCallback _createProviderCallback;
```

Comment:
Any reason this couldn't be shared among all instances? It doesn't look like the callback actually accesses current state?

Reply:
I think we could probably just mark the lambda `static` and inline it where it's used, and everything should just be cached for free.

---
In reply to: [470107965](https://github.com/dotnet/roslyn/pull/46508#discussion_r470107965) [](ancestors = 470107965)

Reply:
I am not comfortable with a static CWT keyed on compilation. In general any static CWT can be evil. Given that allocating this delegate once per instance of the provider is not expensive, I will retain this logic.

Reply:
Oh sorry, I did not see you comment was for just the callback, not the cache. I can definitely make that static.